src/wp-includes/class-wp-theme-json.php

```diff
@@ -767,6 +776,27 @@ public function __construct( $theme_json = array(), $origin = 'theme' ) {
 				}
 			}
 		}
+
+		// In addition to presets, spacingScale (which generates presets) is also keyed by origin.
+		$scale_path    = array( 'settings', 'spacing', 'spacingScale' );
+		$spacing_scale = _wp_array_get( $this->theme_json, $scale_path, null );
+		if ( null !== $spacing_scale ) {
+			// If the spacingScale is not already keyed by origin.
+			if ( empty( array_intersect( array_keys( $spacing_scale ), static::VALID_ORIGINS ) ) ) {
+				_wp_array_set( $this->theme_json, $scale_path, array( $origin => $spacing_scale ) );
+			}
+		}
+
+		// Pre-generate the spacingSizes from spacingScale.
+		$scale_path    = array( 'settings', 'spacing', 'spacingScale', $origin );
+		$spacing_scale = _wp_array_get( $this->theme_json, $scale_path, null );
+		if ( isset( $spacing_scale ) ) {
```

Comment:
Optional:

There was a recent effort to reduce the use of `_wp_array_get`

https://core.trac.wordpress.org/ticket/59405

More verbose, but could also be 

```php
if ( isset( $this->theme_json['settings']['spacing']['spacingScale'][ $origin ] ) ) {
 ....
$spacing_scale_sizes = static::compute_spacing_sizes( $this->theme_json['settings']['spacing']['spacingScale'][ $origin ] );
}
```

might save about `0.000002ms` 😄 

Not a huge deal

Reply:
Using `_wp_array_get` is perfectly valid, IMO. The before/after comparisons I've seen report profiler numbers which are skewed for smaller functions (they report bigger differences than it really is) due to how profilers work (they need to set instrumentation for each function). We'd need before/after comparison with production setup (no profiler, env variables set to production, etc.) to really gauge the impact.

Reply:
Thanks for the awareness of the effort. I learned that we can use the `??` operator now that the minimum PHP version is 7 from the discussion. In many places it looks like the code is more readable and more performant than `_wp_array_get`, and I'll definitely be using that in the future.

However, I'm skeptical of the performance benefit in this particular case considering the alternative is less readable and more prone to errors. Judging from the discussion on https://github.com/WordPress/gutenberg/pull/51116, They were seeing a ~1-2ms improvement (when taking into account the profiler overhead) from replacing more than 2000 `_wp_array_get` calls (https://github.com/WordPress/gutenberg/pull/51116#issuecomment-1603968399).

I think it makes sense to use the null coalescing operator when everything is on one line, but the path here is used in multiple different places, so I'll probably leave this one as-is during my refactors.

Reply:
No worries, thanks for the follow up discussion!